### DISCUSSION TITLE
How to add some typescript logic to Capacitor plugin

### DISCUSSION THREAD

**Discussion by SpertsyanKM** [ionic-team/capacitor#7698] [Q&A]
I’m looking to develop a Capacitor plugin that includes both native iOS and Android components, along with a TypeScript part for mapping and additional logic. From my review of the documentation and examples, it seems that all Capacitor plugins function as intermediaries for native methods, meaning that each time a plugin method is invoked, the call is routed directly to the corresponding native method.

My goal is to intercept these calls on the TypeScript side, perform some processing, invoke the native method myself, handle the response, and then return the result to the user. I attempted to customize the `registerPlugin` call like this:

```typescript
const MySDK = registerPlugin<MySDKPlugin>('MySDK', {
  web: () => import('./web').then(m => new m.MySDKWeb()),
  android: () => import('./mob').then(m => new m.MySDKMob()),
  ios: () => import('./mob').then(m => new m.MySDKMob()),
});
```

However, I found that all calls from Android still go directly to the native part, ignoring the `MySDKMob` class.

If my current approach isn’t correct, I’m considering an alternative: creating an interface for communication with the native code, creating the plugin for internal use without exporting it, and then creating a public interface with my SDK implementation to export as my plugin. However, I haven’t seen any community plugins using this approach, and I’m uncertain if this is the right way to proceed. Is this approach valid, or are there better alternatives to achieve my goal?

**Comment by jcesarmobile** [comment]
We do it for iOS on `text-zoom`, we don't do it on any plugin for Android, but should work the same way.
https://github.com/ionic-team/capacitor-plugins/blob/main/text-zoom/src/index.ts

As for custom tyspescript you can check Google Maps plugin
https://github.com/ionic-team/capacitor-google-maps
